label_studio/data_manager/managers.py

```diff
@@ -509,46 +510,54 @@ def __init__(self, expression, distinct=False, output_field=None, **extra):
         super().__init__(expression, distinct='DISTINCT ' if distinct else '', output_field=output_field, **extra)
 
 
-def annotate_completed_at(queryset):
-    if flag_set('fflag_feat_optic_161_project_settings_for_low_agreement_threshold_score_short', user='auto'):
-        return annotated_completed_at_considering_agreement_threshold(queryset)
-
+def newest_annotation_subquery():
     from tasks.models import Annotation
 
-    newest = Annotation.objects.filter(task=OuterRef('pk')).order_by('-id')[:1]
-    return queryset.annotate(completed_at=Case(When(is_labeled=True, then=Subquery(newest.values('created_at')))))
+    return Annotation.objects.filter(task=OuterRef('pk')).order_by('-id')[:1]
 
 
-def annotated_completed_at_considering_agreement_threshold(queryset):
-    from tasks.models import Annotation
+def base_annotate_completed_at(queryset):
+    newest = newest_annotation_subquery()
+    return queryset.annotate(completed_at=Case(When(is_labeled=True, then=Subquery(newest.values('created_at')))))
+
 
+def annotate_completed_at(queryset):
     LseProject = load_func(settings.LSE_PROJECT)
     get_tasks_agreement_queryset = load_func(settings.GET_TASKS_AGREEMENT_QUERYSET)
 
-    if not queryset.exists():
-        # No tasks to annotate. Quit early
-        return queryset
+    if (
+        get_tasks_agreement_queryset
+        and LseProject
+        and flag_set('fflag_feat_optic_161_project_settings_for_low_agreement_threshold_score_short', user='auto')
+    ):
+        return annotated_completed_at_considering_agreement_threshold(queryset)
 
-    project_id = queryset[0].project_id
+    return base_annotate_completed_at(queryset)
 
-    newest_annotation = Annotation.objects.filter(task=OuterRef('pk')).order_by('-id')[:1]
 
-    if not LseProject:
-        # Not LSE so there will not be agreement_threshold-based task completeness
-        return queryset.annotate(
-            completed_at=Case(When(is_labeled=True, then=Subquery(newest_annotation.values('created_at'))))
-        )
+def annotated_completed_at_considering_agreement_threshold(queryset):
+    LseProject = load_func(settings.LSE_PROJECT)
+    get_tasks_agreement_queryset = load_func(settings.GET_TASKS_AGREEMENT_QUERYSET)
 
-    lse_project = LseProject.objects.filter(project_id=project_id).first()
-    agreement_threshold = lse_project.agreement_threshold if lse_project else None
-    if not (get_tasks_agreement_queryset and agreement_threshold):
-        # This project doesn't use task_agreement so don't consider it when determining completed_at
-        return queryset.annotate(
-            completed_at=Case(When(is_labeled=True, then=Subquery(newest_annotation.values('created_at'))))
+    project_id = queryset.project.id if hasattr(queryset, 'project') and queryset.project is not None else None
```

Comment:
probably it makes sense to return base_annotate_completed_at(queryset) if project_id is None